src/rdkafka_msgset_reader.c

```diff
@@ -1216,3 +1401,230 @@ rd_kafka_msgset_parse (rd_kafka_buf_t *rkbuf,
         return err;
 
 }
+
+
+/**
+ * @brief Offset comparator
+ */
+static int rd_kafka_offset_cmp (const void *_a, const void *_b) {
+        const int64_t *a = _a, *b = _b;
+        return *a - *b;
+}
+
+/**
+ * @brief Pid comparator for rd_kafka_aborted_txn_start_offsets_t
+ */
+static int rd_kafka_aborted_txn_cmp_by_pid (const void *_a, const void *_b) {
+        const rd_kafka_aborted_txn_start_offsets_t *a = _a, *b = _b;
+        return a->pid - b->pid;
+}
+
+
+/**
+ * @brief Free resources associated with an AVL tree node.
+ */
+void rd_kafka_aborted_txn_node_destroy (void *offsets) {
+        rd_list_destroy(&((rd_kafka_aborted_txn_start_offsets_t *)offsets)->offsets);
+}
+
+
+rd_kafka_aborted_txns_t *
+rd_kafka_aborted_txns_new (int32_t txn_cnt) {
+        rd_kafka_aborted_txns_t *aborted_txns;
+        aborted_txns = rd_malloc(sizeof(aborted_txns));
+        aborted_txns->avl = rd_malloc(sizeof(*aborted_txns->avl));
+        rd_avl_init(aborted_txns->avl, rd_kafka_aborted_txn_cmp_by_pid, 0);
+        aborted_txns->list = rd_malloc(sizeof(*aborted_txns->list));
+        rd_list_init(aborted_txns->list, 0,
```

Comment:
can we initialize this list to txn_cnt?

Reply:
yep. it would also then be possible to do away with storing the count separately in rd_kafka_aborted_txns_t, though i've left as is as i think that's semantically better (and I think getting at the list capacity isn't part of the public api).